### ISSUE TITLE
PBSPro: inflated memory readout for Singularity jobs

### ISSUE THREAD

**Issue by ptrebert** [openpbs/openpbs#1554]
Hi,
I have a problem running jobs in Singularity containers due to an inflated memory readout causing the scheduler to kill the job. I have observations from two systems running the same job (same input data, same application container version, same command line):

System 1:
- VM: 36 cores, ~1.4 TB RAM
- CentOS Linux release 7.6.1810 (Core)
- singularity version 3.4.2-1.1.el7
- PBS/TORQUE Version: 6.1.1.1
Several jobs of the same type completed w/o being killed. I once tracked memory consumption for one job/one specific input dataset every minute in a separate script, indicating that the job should complete with a peak consumption of ~600 GB

System 2:
- 72 cores (only 36 or 24 used for testing), ~1.4 TB RAM
- CentOS Linux release 7.7.1908 (Core)
- singularity version 3.5.2
- PBSPro pbs_version = 18.2.5.20190913061152
This system is fully managed, so I only have the info from the local tech support that the job was killed by the scheduler (and not by the cgroup) after requesting > 2 TB RAM. The `qstat` output always indicated a peak of ~1 TB, but I was told that `qstat` is only updating its info every couple of minutes, so it does not tell me much.

System 1 was used for testing, and since all jobs always completed, I did not retain any (qstat) logs. The tech support for System 2 has no idea what's wrong, and it looks like they can't help me much if I don't find an angle where to start here. So, does anybody have any experience with PBSPro plus Singularity, or has even observed this behavior? Thanks in advance.

Best,
Peter

**Comment by agrawalravi90** [comment]
Can you run tracejob on the job id that was killed and post the logs here?

**Comment by ptrebert** [comment]
I can, but the last failed job info is no longer available; getting one of those jobs scheduled for the big machines will take some time

**Comment by agrawalravi90** [comment]
ok, that would help, if PBS killed the job then we should see the reason for it in the logs regardless of where it was running, a container, or an actual machine.

**Comment by ptrebert** [comment]
After an unexpected and long downtime of the cluster where I made this observation, there is finally some progress: based on the help- and insightful comments from the PBSPro support, it looks like the issue was caused by PBSPro enforcing memory limits based on the sum over all RSS values leading to an overestimation of RAM usage (and thus premature kill). My understanding is that this is a configuration issue, i.e., there is no need to have `$enforce mem` set if cgroups are used to watch over job resources.
I am testing the same job again, and it looks like Singularity jobs are still not running smoothly, but this is apparently no longer memory-related.
